pkg/segment/writer/segmetarw.go

```diff
@@ -480,12 +492,107 @@ func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
 	if sfmData.AllPQIDs == nil {
 		sfmData.AllPQIDs = make(map[string]bool)
 	}
-	sfmData.AllPQIDs[newpqid] = true
 
-	// TODO this should be through a channel so that we only have one writer
+	utils.MergeMapsRetainingFirst(sfmData.AllPQIDs, pqidMap)
+
 	writeSfm(segkey, sfmData)
 }
 
+func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
+	BulkBackFillPQSSegmetaEntries(segkey, map[string]bool{newpqid: true})
+}
+
+// AddToBackFillAndEmptyPQSChan adds a new pqid to the channel
+// if emptyPqs is true, then it will also add the EmptyResults for this pqid
+func AddToBackFillAndEmptyPQSChan(segkey string, newpqid string, emptyPqs bool) {
+	pqsChan <- PQSChanMeta{pqid: newpqid, segKey: segkey, emptyPqs: emptyPqs}
+}
+
+func listenBackFillAndEmptyPQSRequests() {
+	// Listen on the channel, every 10 seconds or if the size of the channel is 100,
+	// it would get all the data in the channel and then do the process of Backfilling PQMR files.
+	// This is to avoid multiple writes to the same file
+
+	ticker := time.NewTicker(10 * time.Second) // every 10 seconds
+	defer ticker.Stop()
+
+	buffer := make([]PQSChanMeta, 0, 100)
+	var processing int32
+
+	isProcessing := func() bool {
+		return atomic.LoadInt32(&processing) == 1
+	}
+
+	callProcessBackFillPQSRequests := func() {
+		atomic.StoreInt32(&processing, 1)
+		bufferCopy := make([]PQSChanMeta, len(buffer))
+		copy(bufferCopy, buffer)
+		buffer = buffer[:0]
+		go func(pqsRequests []PQSChanMeta) {
```

Comment:
we shouldn't need a go routine here, once we have accumulated 100 (btw make this a const and not a hard coded value). Then we call some function that is going to take this 100 things and call the process... function